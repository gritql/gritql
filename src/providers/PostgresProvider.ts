import {
  basicSqlDimensions,
  basicSqlMetrics,
  Instruction,
} from '../Instructions/basic'
import { KnexBasedSQLProvider } from './KnexBasedSQLProvider'

export default class PostgresProvider extends KnexBasedSQLProvider {
  name = 'postgres'

  constructor(configuration, connector?) {
    super(configuration, connector || require('postgres'))
  }
  initiateQuery({ table }) {
    const builder = this.getQueryBuilder()
    return { promise: builder.select().from(table), builder }
  }
  instructions = Instructions
  metrics = basicSqlMetrics
  dimensions = basicSqlDimensions

  getMetric(name) {
    const metric = this.metrics.find((instruction) => instruction.name === name)
    return metric
  }
  getDimension(name) {
    const dimension = this.dimensions.find(
      (instruction) => instruction.name === name,
    )
    if (!dimension) return this.dimensions[0] //return default
    return dimension
  }
  getInstruction(name: string) {
    const instruction = this.instructions.find(
      (instruction) => instruction.name === name,
    )
    if (!instruction) {
      throw new Error(`Instruction ${name} not found in ${this.name} provider`)
    }
    return instruction
  }
}

const apply_filters: Instruction = function apply_filters(
  this,
  { promise, builder },
  { args: filters, advancedFilters },
) {
  // todo: refactor, remove advanced filters. All should be normalised if possible.
  if (advancedFilters) {
    if (advancedFilters?.where) {
      promise = promise.where(builder.raw(advancedFilters.where))
    }

    if (advancedFilters?.having) {
      promise = promise.having(builder.raw(advancedFilters.having))
    }
  }
  return filters.reduce((queryPromise, filter, i) => {
    const selector =
      filter[1] === 'in' ? 'whereIn' : i === 0 ? 'where' : 'andWhere'
    return queryPromise[selector].apply(
      queryPromise,
      filter[1] === 'in'
        ? filter.filter((a) => a !== 'in')
        : filter[1] === 'search'
        ? [
            builder.raw(
              `to_tsvector('simple', ??) @@ (plainto_tsquery('simple', ?)::text || ':*')::tsquery`,
              [filter[0], filter[2]],
            ),
          ]
        : filter,
    )
  }, promise)
}

apply_filters.keywords = ['WHERE', 'IN', 'AND', 'OR', 'LIKE', 'ILIKE']
apply_filters.priority = 0

const Instructions: Instruction[] = [apply_filters].concat(basicSqlMetrics)
