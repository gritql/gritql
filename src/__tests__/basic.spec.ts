import gql from 'graphql-tag'
import GritQL from '../gritql'
import PostgresProvider from '../providers/PostgresProvider'

const postgresqlProvider = new PostgresProvider({})
const gritql = new GritQL()
gritql.use(postgresqlProvider)

function getSql(literals: TemplateStringsArray) {
  const defs = gql(literals[0]).definitions
  return gritql
    .queryParser(defs)[0]
    .query.renderQuery()
    .then((context) => {
      return context.promise.toSQL().sql
    })
}

describe('gritql simple query', () => {
  test('generate proper sql', () => {
    getSql`
      query test {
        some: fetch(a: 10, b: 20) {
          troo {
            data: divide(a: t, by: b)
          }
        }
      }
  `.then(matchSnapshot)
  })

  test('basic example works', () => {
    getSql`query table{
    fetch(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
      device {
        date(type: Array) {
          no_baskets: sum(a:no_baskets)
          no_all_baskets: sum(a:no_all_baskets)
          no_unique_products: sum(a:no_unique_products)
          no_brand_products: sum(a:no_brand_products)
          no_uniqie_brand_products: sum(a:no_uniqie_brand_products) 
          total_revenue: sum(a:total_revenue)
          brand_revenue: sum(a:brand_revenue)
        }
      }
    }
  }
  `.then(matchSnapshot)
  })

  test('basic Array with sorting works', () => {
    getSql`query table{
    fetch(category: whatever, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
      channel(type: Array, sort_desc: session_value) {
          session_value: divide(a:revenue, by:sessions)
      }
    }
  }
  `.then(matchSnapshot)
  })

  test('metric functions', () => {
    getSql`query table{
    fetch(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
      device {
        date(type: Array) {
          no_unique_products: sum(a:no_unique_products)
          no_brand_products: sum(a:no_brand_products)
          average: divide(a:no_baskets, by:no_all_baskets)
        }
      }
    }
  }
  `.then(matchSnapshot)
  })
})

function matchSnapshot(sql) {
  expect(sql).toMatchSnapshot()
}
