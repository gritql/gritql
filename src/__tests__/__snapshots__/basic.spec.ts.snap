// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gritql simple query basic example works 1`] = `"select "device", "date", sum("no_baskets"), sum("no_all_baskets"), sum("no_unique_products"), sum("no_brand_products"), sum("no_uniqie_brand_products"), sum("total_revenue"), sum("brand_revenue") from "table" where "table"."brand" = ? and "table"."country" = ? and "table"."date" > ? and "table"."date" < ? group by "device", "date""`;

exports[`gritql simple query generate proper sql 1`] = `"select "troo", SUM("t")::float4//NULLIF(SUM("b")::float4, 0) AS "data" from "test" where "test"."a" = ? and "test"."b" = ? group by "troo""`;
